relay/circuit-v2.md

```diff
@@ -0,0 +1,398 @@
+R -> A: [hop] STATUS:OK
+
+hnote over A: Reservation timeout approaching.
+hnote over A: Refresh.
+
+A -> R: [hop] RESERVE
+R -> A: [hop] STATUS:OK
+
+hnote over A: ...
+
+== Circuit Establishment ==
+
+B -> R: [hop] CONNECT to A
+R -> A: [stop] CONNECT from B
+A -> R: [stop] STATUS:OK
+R -> B: [hop] STATUS:OK
+
+B <-> A: Connection
+@enduml
+```
+  
+</details>
+
+
+The first part of the interaction is _A_'s reservation of a relay slot
+in _R_.  This is accomplished by opening a connection to _R_ and
+sending a `RESERVE` message in the `hop` protocol; if the reservation
+is successful, the relay responds with a `STATUS:OK` message and
+provides _A_ with a reservation voucher. _A_ keeps the connection to _R_ alive for the duration of the reservation, refreshing the reservation as needed.
+
+The second part of the interaction is the establishment of a circuit
+switch connection from _B_ to _A_ through _R_.  It is assumed that _B_
+has obtained a circuit multiaddr for _A_ of the form
+`/p2p/QmR/p2p-circuit/p2p/QmA` out of band using some peer discovery
+service (eg. the DHT or a rendezvous point).
+
+In order to connect to _A_, _B_ then conncts to _R_, opens a `hop`
+protocol stream and sends a `CONNECT` message to the relay.  The relay
+verifies that it has a reservation and connection for _A_ and opens a
+`stop` protocol stream to _A_, sending a `CONNECT` message.
+
+Peer _A_ then responds to the relay with a `STATUS:OK` message, which
+responds to _B_ with a `STATUS:OK` message in the open `hop` stream
+and then proceeds to bridge the two streams into a relayed connection.
+The relayed connection flows in the `hop` stream between the
+connection initiator and the relay and in the `stop` stream between
+the relay and the connection termination point.
+
+### Hop Protocol
+
+The Hop protocol governs interaction between clients and the relay;
+it uses the protocol ID `/libp2p/circuit/relay/0.2.0/hop`.
+
+There are two parts of the protocol:
+- reservation, by peers that wish to receive relay service
+- connection initiation, by peers that wish to connect to a peer through the relay.
+
+#### Reservation
+
+In order to make a reservation, a peer opens a connection to the relay and sends a `HopMessage` with
+`type = RESERVE`:
+
+```
+HopMessage {
+  type = RESERVE
+}
+```
+
+The relay responds with a `HopMessage` of `type = STATUS`, indicating whether the reservation has
+been accepted.
+
+If the reservation is accepted, then the message has the following form:
+```
+HopMessage {
+  type = STATUS
+  status = OK
+  reservation = Reservation {...}
+  limit = Limit {...}
+}
+```
+
+The `reservation` field provides information about the reservation itself; the struct has the following fields:
+```
+Reservation {
+   expire = ...
+   addrs = [...]
+   voucher = ...
+}
+```
+
+- the `expire` field contains the expiration time as a UTC UNIX time. The reservation becomes invalid after this time and it's the responsibility of the client to refresh.
```

Comment:
Is it up to the server to drop the connection if the client fails to refresh after a specific grace period?

Reply:
the server may or may not drop a connection according to its connection management policy; the expectation is that the server will make a best effort attempt to maintain the connection for the duration of the reservation.

Perhaps we should add some wording about this.

Reply:
added wording.